### PR TITLE
Show tasks only for lecturers/tutors

### PR DESCRIPTION
**Pull request by timst01**
I did not change getTaskResult because i think the result is only accessed trough the Task, so if the Task is not visible the results can not be accessed

### CODE REVIEW

**Review by mxsph — CHANGES_REQUESTED** [review]
Some of the commits should not be in this PR as they belong to the PR 
#754. You could remove them for example with an ineractive rebase (see https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_changing_multiple) or cherry-pick (see https://git-scm.com/docs/git-cherry-pick).

**Comment by pdrtzk** [comment]
The commits that do not belong to the PR should be removed now.

**Review by pdrtzk — CHANGES_REQUESTED** [review]
- [ ] Update the [api documentation](https://github.com/thm-mni-ii/feedbacksystem/blob/dev/modules/fbs-core/api/api-docs.yml).
- [ ] Instead of checking if the task `isPublic` check if it `isPrivate`, which should be default `false`.
- [ ] In the api requests `isPrivate` should be optional, so that frontend still works.

**Review by pdrtzk — CHANGES_REQUESTED** [review]
There is currently a bug with the authorization caused by this line:
https://github.com/thm-mni-ii/feedbacksystem/blob/991614684efd5eef05339d0bcb73f6ab09716b67/modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala#L72

Because the `getOne()` function in filter is called on each object in the stream, the authorization inside of `getOne()` is called too often.

You can outsource the following code out of `getOne()` and inside a new function. This must be called instead of the code then and also in the filter.

https://github.com/thm-mni-ii/feedbacksystem/blob/991614684efd5eef05339d0bcb73f6ab09716b67/modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala#L114-L134

**Review by pdrtzk — APPROVED** [review]
Works now as expected.  👍

**Review by mxsph — CHANGES_REQUESTED** [review]
- [ ] If a task in a course is `private`, no tasks are displayed for a student. Because the `/{cid}/tasks/results` then throws a `403` error.
- [ ] Private assignments should not only be returned to administrators, but also to "course" lecturers.

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala [hunk 1]
```diff
@@ -49,8 +50,13 @@ class TaskController {
   @GetMapping(value = Array("/{cid}/tasks"))
   @ResponseBody
   def getAll(@PathVariable("cid") cid: Int, req: HttpServletRequest, res: HttpServletResponse): List[Task] = {
-    authService.authorize(req, res)
-    taskService.getAll(cid)
+    val auth = authService.authorize(req, res)
+    val someCourseRole = courseRegistration.getParticipants(cid).find(_.user.id == auth.id).map(_.role)
```
Comment (mxsph) [inline modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala:54]: You should improve this code as it always has to search all users in the course. For this I would add a new method in the `courseRegistration` service which returns the role of a user in the course (for this you can orientate yourself on the function `getCoursePrivileges`).

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala [hunk 2]
```diff
@@ -65,7 +71,29 @@ class TaskController {
   @ResponseBody
   def getTaskResults(@PathVariable("cid") cid: Int, req: HttpServletRequest, res: HttpServletResponse): Seq[UserTaskResult] = {
     val auth = authService.authorize(req, res)
-    taskService.getTaskResults(cid, auth.id)
+    val someCourseRole = courseRegistration.getParticipants(cid).find(_.user.id == auth.id).map(_.role)
```
Comment (mxsph) [inline modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala:74]: You should improve this code as it always has to search all users in the course. For this I would add a new method in the `courseRegistration` service which returns the role of a user in the course (for this you can orientate yourself on the function `getCoursePrivileges`).

File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala [hunk 3]
```diff
@@ -65,7 +71,29 @@ class TaskController {
...
+    * @return
+    */
+  def isTaskPublic(cid: Int, taskId: Int, auth: User): Boolean = {
+    try {
+      getTaskById(cid, taskId, auth)
+      true
+    } catch {
+      case _ => false
+    }
   }
```
Comment (mxsph) [inline modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/controller/TaskController.scala:97]: I would solve the problem a bit differently and not use a `try catch` but pass a parameter to the method `getTaskById` which determines whether an exception is thrown or not.

**Review by mxsph — CHANGES_REQUESTED** [review]
- [x] It should not be possible for a user to make a submission to a private task.
- [ ] It should not be possible for a user to retrieve results from a private task.

**Review by mxsph — APPROVED** [review]

**Review by Zitrone44 — APPROVED** [review]

**Pull request merged**